### ISSUE TITLE
wrong shrink_source calculation in draw_networkx_edges

### ISSUE THREAD

**Issue by sanghack81** [networkx/networkx#3805]
I would like to talk about a problem with drawing edges for `DiGraph` (`draw_networkx_edges`).

`shrink_source` is treated differently from `shrink_target`.
It would be natural to think that there is no arrow at the source since an arrow is drawn from a source to a target!
However, one may want to use < - - > bidirected dashed arrows, for example.
If `shrink_source` is not treated the same as `shrink_target`, the arrow tip at the source will be buried under the source node.

Hence, the following modification is suggested:
```python
shrink_source = 0  # space from source to tail
shrink_target = 0  # space from  head to target

if np.iterable(node_size):  # many node sizes
    src_node, dst_node = edgelist[i][:2]
    shrink_source = to_marker_edge(node_size[nodelist.index(src_node)], node_shape)
    shrink_target = to_marker_edge(node_size[nodelist.index(dst_node)], node_shape)
else:
    shrink_source = shrink_target = to_marker_edge(node_size, node_shape)

if shrink_source < min_source_margin:
    shrink_source = min_source_margin

if shrink_target < min_target_margin:
    shrink_target = min_target_margin
```

Then, the modified code can yield a correctly-styled graph.
```python
import networkx as nx
g = nx.DiGraph()
g.add_nodes_from({'X','Y'})
g.add_edges_from([('X','Y')])

pos = nx.nx_pydot.pydot_layout(g, prog='dot')
nx.draw_networkx_nodes(g, pos, node_color='#ffffff', edgecolors='#000000')
nx.draw_networkx_edges(g, pos)
nx.draw_networkx_labels(g, pos)

g2 = nx.DiGraph()
g2.add_nodes_from({'X','Y'})
g2.add_edges_from([('X','Y')])
nx.draw_networkx_edges(g2, pos, style='dashed', arrowstyle='<|-|>', arrows=True, connectionstyle=ConnectionStyle.Arc3(rad=0.5))

g3 = nx.DiGraph()
g3.add_nodes_from({'X','Y'})
g3.add_edges_from([('X','Y')])
nx.draw_networkx_edges(g3, pos, style='dotted', arrowstyle='-|>', arrows=True, connectionstyle=ConnectionStyle.Arc3(rad=-0.5))
plt.show()
```
<img width="164" alt="Screen Shot 2020-01-30 at 11 52 10 PM" src="https://user-images.githubusercontent.com/4041206/73513460-8f479580-43bb-11ea-9f08-cef77ead26fe.png">

**Comment by sanghack81** [comment]
#3747 already addressed `linestyle`

**Comment by dschult** [comment]
This sounds like good proposal.  Thanks you!

**Comment by dschult** [comment]
closed by #3806
